### PR TITLE
isl: Fix isl 0.20 on riscv due to out of date config.guess

### LINKED ISSUE #226578: Build failure: isl_0_20
### Steps To Reproduce

1. build isl-0.20 on riscv64-linux

### Build log

```
building '/nix/store/q7c67kcd89yymn222igjz58dlgaij489-isl-0.20.drv'...
unpacking sources
unpacking source archive /nix/store/x9v0hmr76dm68pijzldzxks01iqzag1z-isl-0.20.tar.xz
source root is isl-0.20
setting SOURCE_DATE_EPOCH to timestamp 1532811285 of file isl-0.20/interface/isl.py
patching sources
configuring
fixing libtool script ./ltmain.sh
./configure
configure flags: --disable-static --disable-dependency-tracking --prefix=/nix/store/9hkiw8kgbi3zcclxk3l8jh624gin57pj-isl-0.20 --with-gcc-arch=generic
checking for a BSD-compatible install... /nix/store/am27aaylz23yrws5bl5v858bg57qjdg3-bootstrap-tools/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/am27aaylz23yrws5bl5v858bg57qjdg3-bootstrap-tools/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
./configure: line 3487: cd: /nix/store/9hkiw8kgbi3zcclxk3l8jh624gin57pj-isl-0.20: No such file or directory
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... none
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... none
checking for C compiler vendor... gnu
checking build system type... ./config.guess: unable to guess system type

This script, last modified 2014-03-23, has failed to recognize
the operating system you are using. It is advised that you
download the most up to date version of the config scripts from

  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=HEAD
and
  http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.sub;hb=HEAD

If the version you run (./config.guess) is already up to date, please
send the following data and any information you think might be
pertinent to <config-patches@gnu.org> in order to provide the needed
information to handle your system.

config.guess timestamp = 2014-03-23

uname -m = riscv64
uname -r = 5.15.0
uname -s = Linux
uname -v = #1-NixOS SMP Mon Mar 20 17:53:42 UTC 2023

/usr/bin/uname -p =
/bin/uname -X     =

hostinfo               =
/bin/universe          =
/usr/bin/arch -k       =
/bin/arch              =
/usr/bin/oslevel       =
/usr/convex/getsysinfo =

UNAME_MACHINE = riscv64
UNAME_RELEASE = 5.15.0
UNAME_SYSTEM  = Linux
UNAME_VERSION = #1-NixOS SMP Mon Mar 20 17:53:42 UTC 2023
configure: error: cannot guess build type; you must specify one
error: builder for '/nix/store/q7c67kcd89yymn222igjz58dlgaij489-isl-0.20.drv' failed with exit code 1

```

### Additional context

isl 0.20 is ancient, and needs a config.guess update I guess

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
(fails to run due to above error)
```

### PR DESCRIPTION
**Pull request by kangaroo**
###### Description of changes

Add autoreconfHook to isl.  Fixes #226578.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by NickCao — APPROVED** [review]
diff LGTM, did not test.

**Review by ghost — APPROVED** [review]
@kangaroo, thank you for taking the time (and doing the rebuilds) to track this down and fixing it.  This is the right solution for this kind of problem.

@vcunat, is it okay to merge this to `staging`?

Built on:
- [x] `x86_64-linux`
- [x] `powerpc64le-linux`
- [x] `aarch64-linux`
- [x] `mips64el-linux` (cross from `x86_64-linux`)

**Comment by Artturin** [comment]
commit msg should follow CONTRIBUTING.md

**Comment by kangaroo** [comment]
Commit msg was updated

**Comment by vcunat** [comment]
If I have any concerns, it's about the impact on the "length" of our linux stdenv bootstrapping.

**Review by leifhelm — APPROVED** [review]

**Comment by kangaroo** [comment]
Can we merge this?

**Comment by amjoseph-nixpkgs** [comment]
> Can we merge this?

It may have to wait until after the 23.05 release.

This PR triggers a rebuild of every single package in the entirety of nixpkgs.  That is a massive amount of recompiling in order to fix a problem on a platform for which Hydra does not generate binaries.

**Comment by vcunat** [comment]
That's not an issue.  `stdenv` will be rebuilt regardless due to other changes.

**Comment by vcunat** [comment]
A simple option might be to add the dependency optionally, only on platforms where it might be needed.  (or at least exclude platforms where we care about bootstrap speed)

**Comment by amjoseph-nixpkgs** [comment]
Yeah this could get ugly.  `autoconf` needs `gettext`, which takes an eternity to compile (serial build process, one core used at most).

Before this PR nothing needs autoconf until we get to `gawk` (i.e. building our own, not the one from the bootstrapFiles) or `glibc`'s special `python3` -- both of those are after `xgcc`, but `isl` is before.

So this would add a completely new copy of: `autoconf`, `libtool`, `automake`, `gettext`, `m4`, `zlib`, and `texinfo`.



> A simple option might be to add the dependency optionally, only on platforms where it might be needed. (or at least exclude platforms where we care about bootstrap speed)

Yeah I think that's the way to go.

File: pkgs/development/libraries/isl/generic.nix [hunk 1]
```diff
@@ -19,6 +19,7 @@ stdenv.mkDerivation {
   inherit patches;
 
   strictDeps = true;
+  nativeBuildInputs = [ autoreconfHook ];
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/libraries/isl/generic.nix:22]:
```suggestion
  # Note: isl is used very early in the bootstrap process, so including
  # autoreconfHook here as a dependency triggers a lot of extra builds.
  # So we include it only where needed (old autoconf did not understand RiscV)
  nativeBuildInputs = lib.optionals stdenv.hostPlatform.isRiscV && lib.versionOlder version "0.24" [ autoreconfHook ];
```

File: pkgs/development/libraries/isl/generic.nix [hunk 2]
```diff
@@ -19,6 +19,10 @@ stdenv.mkDerivation {
   inherit patches;
 
   strictDeps = true;
+  # Note: isl is used very early in the bootstrap process, so including
+  # autoreconfHook here as a dependency triggers a lot of extra builds.
+  # So we include it only where needed (old autoconf did not understand RiscV)
+  nativeBuildInputs = lib.optionals stdenv.hostPlatform.isRiscV && lib.versionOlder version "0.24" [ autoreconfHook ];
```
Comment (vcunat) [inline pkgs/development/libraries/isl/generic.nix:25]: The conditional needs parentheses, as noticed by OfBorg.

**Review by vcunat — CHANGES_REQUESTED** [review]

**Review by 0x4A6F — APPROVED** [review]
Built on:
- [x] `riscv64-linux` (native build took 10h54m23s on VisionFive 2)

File: pkgs/development/libraries/isl/generic.nix [hunk 3]
```diff
@@ -19,6 +19,7 @@ stdenv.mkDerivation {
   inherit patches;
 
   strictDeps = true;
+  nativeBuildInputs = lib.optionals (stdenv.hostPlatform.isRiscV && lib.versionOlder version "0.24") [ autoreconfHook ];
```
[the remark this answers is not in the document]
Comment (vcunat) [reply pkgs/development/libraries/isl/generic.nix:22]: I don't think it's economical in this case.  Lots of other riscv breakages can be introduced and not found until built for riscv.

**Pull request merged**